From 7f0d456c9e901934bce8358dfe8faab492058323 Mon Sep 17 00:00:00 2001
From: David Ward <david.ward@ll.mit.edu>
Date: Sun, 5 Apr 2026 15:33:16 -0400
Subject: [PATCH 3/5] lib: Remove parameter from wrapper for lyd_new_path2()

libyang 5.1.0 changed lyd_new_path2(), replacing a parameter of type
LYD_ANYDATA_VALUETYPE with a different parameter of type uint32_t.
The former was always set to 0 by FRR, and the latter can be set to 0
as well; but remove this parameter from the wrapper to avoid issues.

Also fix a variable name in the wrapper: the value size is in bytes.

Signed-off-by: David Ward <david.ward@ll.mit.edu>
Signed-off-by: Daniel Baumann <mail@daniel-baumann.ch>
---
 lib/mgmt_be_client.c    |  4 ++--
 lib/northbound.c        |  5 ++---
 lib/northbound_cli.c    |  4 ++--
 lib/northbound_grpc.cpp |  6 ++----
 lib/northbound_oper.c   |  2 +-
 lib/yang.c              | 15 +++++++--------
 lib/yang.h              |  3 +--
 7 files changed, 17 insertions(+), 22 deletions(-)

diff --git a/lib/mgmt_be_client.c b/lib/mgmt_be_client.c
index 42b4ce974f..df9cc471d4 100644
--- a/lib/mgmt_be_client.c
+++ b/lib/mgmt_be_client.c
@@ -851,7 +851,7 @@ static void be_client_handle_rpc(struct mgmt_be_client *client, uint64_t txn_id,
 		 * It is especially needed for actions, because their parents
 		 * may hold necessary information.
 		 */
-		err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, 0, NULL, &input);
+		err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, NULL, &input);
 	}
 	if (err) {
 		be_client_send_error(client, txn_id, rpc_msg->req_id, false, -EINVAL,
@@ -859,7 +859,7 @@ static void be_client_handle_rpc(struct mgmt_be_client *client, uint64_t txn_id,
 		return;
 	}
 
-	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, 0, NULL, &output);
+	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, NULL, &output);
 	if (err) {
 		lyd_free_all(input);
 		be_client_send_error(client, txn_id, rpc_msg->req_id, false,
diff --git a/lib/northbound.c b/lib/northbound.c
index ac004d27c6..6bfa8b2042 100644
--- a/lib/northbound.c
+++ b/lib/northbound.c
@@ -738,7 +738,7 @@ static LY_ERR dnode_create(struct nb_config *candidate, const char *xpath, const
 	struct lyd_node *dnode = NULL;
 	LY_ERR err;
 
-	err = yang_new_path2(candidate->dnode, ly_native_ctx, xpath, value, 0, 0, options, NULL,
+	err = yang_new_path2(candidate->dnode, ly_native_ctx, xpath, value, 0, options, NULL,
 			     &dnode);
 	if (err) {
 		flog_warn(EC_LIB_LIBYANG, "%s: lyd_new_path(%s) failed: %d",
@@ -907,8 +907,7 @@ static int nb_candidate_edit_tree_add(struct nb_config *candidate,
 	 * merged later with candidate.
 	 */
 	if (parent_xpath) {
-		err = yang_new_path2(NULL, ly_native_ctx, parent_xpath, NULL, 0, 0, 0, &tree,
-				     &parent);
+		err = yang_new_path2(NULL, ly_native_ctx, parent_xpath, NULL, 0, 0, &tree, &parent);
 		if (err) {
 			yang_print_errors(ly_native_ctx, errmsg, errmsg_len);
 			ret = NB_ERR;
diff --git a/lib/northbound_cli.c b/lib/northbound_cli.c
index 98ac568f72..b8ff194e07 100644
--- a/lib/northbound_cli.c
+++ b/lib/northbound_cli.c
@@ -298,7 +298,7 @@ int nb_cli_rpc(struct vty *vty, const char *xpath, struct lyd_node **output_p)
 	}
 
 	/* create input tree */
-	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, 0, NULL, &input);
+	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, NULL, &input);
 	assert(err == LY_SUCCESS);
 
 	for (size_t i = 0; i < vty->num_rpc_params; i++) {
@@ -319,7 +319,7 @@ int nb_cli_rpc(struct vty *vty, const char *xpath, struct lyd_node **output_p)
 	assert(err == LY_SUCCESS);
 
 	/* create output tree root for population in the callback */
-	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, 0, NULL, &output);
+	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, NULL, &output);
 	assert(err == LY_SUCCESS);
 
 	ret = nb_callback_rpc(nb_node, xpath, input, output, errmsg,
diff --git a/lib/northbound_grpc.cpp b/lib/northbound_grpc.cpp
index 97d5418a90..25ef9fa9ac 100644
--- a/lib/northbound_grpc.cpp
+++ b/lib/northbound_grpc.cpp
@@ -1032,8 +1032,7 @@ grpc::Status HandleUnaryExecute(
 				    "Unknown data path");
 
 	// Create input data tree.
-	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, (LYD_ANYDATA_VALUETYPE)0, 0,
-			     NULL, &input_tree);
+	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, NULL, &input_tree);
 	if (err != LY_SUCCESS) {
 		return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
 				    "Invalid data path");
@@ -1061,8 +1060,7 @@ grpc::Status HandleUnaryExecute(
 	}
 
 	// Create output data tree.
-	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, (LYD_ANYDATA_VALUETYPE)0, 0,
-			     NULL, &output_tree);
+	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, NULL, &output_tree);
 	if (err != LY_SUCCESS) {
 		lyd_free_tree(input_tree);
 		return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
diff --git a/lib/northbound_oper.c b/lib/northbound_oper.c
index 0ab1600183..f10f112596 100644
--- a/lib/northbound_oper.c
+++ b/lib/northbound_oper.c
@@ -441,7 +441,7 @@ static enum nb_error nb_op_xpath_to_trunk(const char *xpath_in, char **xpath_out
 
 	ly_temp_log_options(&llopts);
 	for (;;) {
-		err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, LYD_NEW_PATH_UPDATE,
+		err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, LYD_NEW_PATH_UPDATE,
 				     NULL, trunk);
 		if (err == LY_SUCCESS)
 			break;
diff --git a/lib/yang.c b/lib/yang.c
index 66f722f7b0..40dce87b35 100644
--- a/lib/yang.c
+++ b/lib/yang.c
@@ -707,7 +707,7 @@ struct lyd_node *yang_state_new(struct lyd_node *tree, const char *path, const c
 	struct lyd_node *dnode, *parent;
 	LY_ERR err;
 
-	err = yang_new_path2(tree, ly_native_ctx, path, value, 0, 0, LYD_NEW_PATH_UPDATE, &parent,
+	err = yang_new_path2(tree, ly_native_ctx, path, value, 0, LYD_NEW_PATH_UPDATE, &parent,
 			     &dnode);
 	assert(err == LY_SUCCESS);
 
@@ -916,7 +916,7 @@ LY_ERR yang_parse_data(const char *xpath, LYD_FORMAT format, bool as_subtree, bo
 		 * a common YANG JSON technique (vs XML which starts all
 		 * data trees from the root).
 		 */
-		err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, 0, &parent, &subtree);
+		err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, &parent, &subtree);
 		if (err != LY_SUCCESS)
 			goto done;
 		err = lyd_find_path(parent, xpath, false, &subtree);
@@ -1016,7 +1016,7 @@ LY_ERR yang_parse_restconf_rpc(const char *xpath, LYD_FORMAT format, const char
 		return LY_ENOTFOUND;
 	}
 	/* Get the tree for the RPC/Action */
-	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, 0, NULL, &dnode);
+	err = yang_new_path2(NULL, ly_native_ctx, xpath, NULL, 0, 0, NULL, &dnode);
 	if (err) {
 		zlog_err("Failed to create parent node for action: %s", ly_last_errmsg());
 		goto done;
@@ -1070,8 +1070,7 @@ LY_ERR yang_parse_rpc(const char *xpath, LYD_FORMAT format, const char *data, bo
 			return LY_EINVAL;
 		}
 
-		err = yang_new_path2(NULL, ly_native_ctx, parent_xpath, NULL, 0, 0, 0, NULL,
-				     &parent);
+		err = yang_new_path2(NULL, ly_native_ctx, parent_xpath, NULL, 0, 0, NULL, &parent);
 		XFREE(MTYPE_TMP, parent_xpath);
 		if (err) {
 			zlog_err("Failed to create parent node for action: %s",
@@ -1733,10 +1732,10 @@ LY_ERR yang_new_term_bin(struct lyd_node *parent, const struct lys_module *modul
 }
 
 LY_ERR yang_new_path2(struct lyd_node *parent, const struct ly_ctx *ctx, const char *path,
-		      const void *value, uint32_t size, LYD_ANYDATA_VALUETYPE value_type,
-		      uint32_t options, struct lyd_node **new_parent, struct lyd_node **new_node)
+		      const void *value, uint32_t value_size_bytes, uint32_t options,
+		      struct lyd_node **new_parent, struct lyd_node **new_node)
 {
-	return lyd_new_path2(parent, ctx, path, value, LY_SZ(size), value_type, options,
+	return lyd_new_path2(parent, ctx, path, value, LY_SZ(value_size_bytes), 0, options,
 			     new_parent, new_node);
 }
 
diff --git a/lib/yang.h b/lib/yang.h
index 2764647d5f..dac6c43239 100644
--- a/lib/yang.h
+++ b/lib/yang.h
@@ -958,8 +958,7 @@ extern LY_ERR yang_new_term_bin(struct lyd_node *parent, const struct lys_module
 	_Pragma("GCC error \"Use yang_new_term_bin() instead of lyd_new_term_bin()\"")
 
 extern LY_ERR yang_new_path2(struct lyd_node *parent, const struct ly_ctx *ctx, const char *path,
-			     const void *value, uint32_t value_size_bits,
-			     LYD_ANYDATA_VALUETYPE value_type, uint32_t options,
+			     const void *value, uint32_t value_size_bytes, uint32_t options,
 			     struct lyd_node **new_parent, struct lyd_node **new_node);
 #define lyd_new_path2(...) _Pragma("GCC error \"Use yang_new_path2() instead of lyd_new_path2()\"")
 
-- 
2.55.0

